tls/extensions/s2n_server_key_share.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/extensions/s2n_server_key_share.h"
+
+#include "tls/s2n_client_extensions.h"
+#include "utils/s2n_safety.h"
+
+static int s2n_extensions_server_key_share_send_check(struct s2n_connection *conn);
+
+/*
+ * Check whether client has sent a corresponding curve and key_share
+ */
+static int s2n_extensions_server_key_share_send_check(struct s2n_connection *conn)
+{
+    const struct s2n_ecc_named_curve *server_curve, *client_curve;
+    server_curve = conn->secure.server_ecc_params.negotiated_curve;
+    notnull_check(server_curve);
+
+    int curve_index = -1;
+    for (int i = 0; i < S2N_ECC_SUPPORTED_CURVES_COUNT; i++) {
+        if (server_curve == &s2n_ecc_supported_curves[i]) {
+            curve_index = i;
+            break;
+        }
+    }
+    gt_check(curve_index, -1);
+
+    const struct s2n_ecc_params client_ecc = conn->secure.client_ecc_params[curve_index];
+    client_curve = client_ecc.negotiated_curve;
+
+    S2N_ERROR_IF(client_curve == NULL, S2N_ERR_BAD_KEY_SHARE);
+    S2N_ERROR_IF(client_curve != server_curve, S2N_ERR_BAD_KEY_SHARE);
+    S2N_ERROR_IF(client_ecc.ec_key == NULL, S2N_ERR_BAD_KEY_SHARE);
+
+    return 0;
+}
+
+/*
+ * Calculate the data length for Server Key Share extension
+ * based on negotiated_curve selected in server_ecc_params.
+ *
+ * This functions does not error, but s2n_extensions_server_key_share_send() would
+ */
+int s2n_extensions_server_key_share_send_size(struct s2n_connection *conn)
+{
+    const struct s2n_ecc_named_curve* curve = conn->secure.server_ecc_params.negotiated_curve;
+
+    if (curve == NULL) {
+        return 0;
+    }
+
+    const int key_share_size = S2N_SIZE_OF_EXTENSION_TYPE
+        + S2N_SIZE_OF_EXTENSION_DATA_SIZE
+        + S2N_SIZE_OF_NAMED_GROUP
+        + S2N_SIZE_OF_KEY_SHARE_SIZE
+        + curve->share_size;
+
+    return key_share_size;
+}
+
+/*
+ * Sends Key Share extension in Server Hello.
+ *
+ * Expects negotiated_curve to be set and generates a ephemeral key for key sharing
+ */
+int s2n_extensions_server_key_share_send(struct s2n_connection *conn, struct s2n_stuffer *out)
+{
+    GUARD(s2n_extensions_server_key_share_send_check(conn));
```

Comment:
Prior to calling this function, could we verify that  `conn->secure.server_ecc_params.negotiated_curve` is set, and if not set, error out?

Reply:
> Prior to calling this function

Are you referring to when `s2n_extensions_server_key_share_send_size()` is called, or even before `s2n_server_extensions_send()` gets called?

My preference is that key checks/selection should be done before `s2n_server_extensions_send()`

Reply:
Having the key checks verified prior to `s2n_server_extensions_send()` seems good to me.